### ISSUE TITLE
Hardware accelerated AES for ARM

### ISSUE THREAD

**Issue by newpavlov** [RustCrypto/block-ciphers#10]
Similar to `aesni` crate it would be nice to have hardware accelerated AES support for ARM using specialized [instructions](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0514g/way1395175472464.html).

**Comment by quininer** [comment]
Should we plan to use stdsimd? https://github.com/rust-lang-nursery/stdsimd/issues/295

**Comment by newpavlov** [comment]
Excellent! I've subscribed to both issues and will try to implement both ARM and x86 versions once they get implemented in the `stdsimd`. (for x86 I [have](https://github.com/RustCrypto/block-ciphers/tree/llvmint_aesni) LLVM intrinsics based draft) One thing which I would like to have is an ability to specify `target_feature` in the `Cargo.toml`, otherwise intrinsics based solution will be less convenient to use, as you'll have to list several features in the `RUSTFLAGS`.

**Comment by jasondavies** [comment]
FYI, I recently added aarch64 AES intrinsics in rust-lang-nursery/stdsimd#398.  I believe AES intrinsics for x86 are in there already.

**Comment by newpavlov** [comment]
Yes, I've seen it, thank you for this addition! When I'll have time I'll try to play with them and implement the crate. (if no one beats me to it that is)

AES-NI intrinsics already utilized in the `aesni` crate, as for Intel  SHA extensions currently I don't have an appropriate CPU to test code on.

**Comment by tarcieri** [comment]
@jack-signal I saw you have hardware accelerated `aarch64` implemented here:

https://github.com/signalapp/libsignal-client/blob/ 946e067/rust/aes-gcm-siv/src/aes/aarch64.rs

Any interest in upstreaming that?

**Comment by jack-signal** [comment]
At this point everything Signal is putting out, including that crate, is AGPL, I doubt there is interest in relicensing so probably wouldn't work. It's worth pointing out though that the linked code is just for AES-256 in forward direction because that is all we needed, and it doesn't implement any of the RustCrypto traits so a ways off from something you could use.

Someone recently pointed me at the MIT licensed https://github.com/shadowsocks/crypto2/blob/master/src/blockcipher/aes/aarch64.rs which might work better as a starting point. But that code doesn't do instruction pipelining, and the key schedule isn't constant time :(

**Comment by tarcieri** [comment]
Ugh, alright, it seems both those approaches are a no-go.

**Comment by tarcieri** [comment]
@newpavlov were you thinking of trying to do an implementation of this from scratch?

**Comment by tarcieri** [comment]
SUPERCOP contains a public domain implementation of the full AES-GCM including one based on ARMv9 intrinsics which works on 8-blocks-at-a-time in parallel:

https://github.com/floodyberry/supercop/blob/master/crypto_aead/aes256gcmv1/dolbeau/armv8crypto/armv8crypto.c

**Comment by newpavlov** [comment]
@tarcieri 
Unfortunately I don't have an easy way to test code which uses ARM crypto extension. Unfortunately RPi4 does not support it and while my phone probably has it, right now I don't have development environment and necessary knowledge for it.

**Comment by tarcieri** [comment]
@newpavlov no worries!

I have an Apple M1 Mac Mini that as of yesterday now has the beta Rust compiler I can test things on. I might experiment to see if I can get it working.

**Comment by myfreeweb** [comment]
@newpavlov you can always get an [AWS EC2 instance](https://aws.amazon.com/ec2/graviton/) to test code, until the end of 2020 one `t4g.micro` is available for free for all accounts

**Comment by jack-signal** [comment]
qemu supports the Aarch64 crypto extensions and is pretty easy to integrate into `cargo test`, just install qemu and aarch64 cross gcc, and add to `~/.cargo/config`

```
[target.aarch64-unknown-linux-gnu]
linker = "aarch64-linux-gnu-gcc"
runner = ["qemu-aarch64", "-L", "/usr/aarch64-linux-gnu"]
```

Then you can build and test Aarch64 code on x86-64

**Comment by tarcieri** [comment]
As it were, we already use `cross` to test against aarch64:

https://github.com/RustCrypto/block-ciphers/blob/master/.github/workflows/aes.yml#L117-L147

**Comment by tarcieri** [comment]
I landed ARMv8 Cryptography Extensions support in #250, however it's only tested on and therefore presently gated on `aarch64` targets.

If you're interested in support for 32-bit ARMv8 targets or other types of ARM acceleration, please leave a comment on this issue or make a new issue.
